src/librustc_codegen_ssa/back/write.rs

```diff
@@ -380,26 +371,42 @@ pub fn start_async_codegen<B: ExtraBackendMethods>(
         allocator_config.emit_bc = true;
     }
 
-    // Emit compressed bitcode files for the crate if we're emitting an rlib.
-    // Whenever an rlib is created, the bitcode is inserted into the archive in
-    // order to allow LTO against it.
-    if need_crate_bitcode_for_rlib(sess) {
-        modules_config.emit_bc_compressed = true;
-        allocator_config.emit_bc_compressed = true;
-    }
-
     let emit_obj =
         if sess.target.target.options.obj_is_bitcode || sess.opts.cg.linker_plugin_lto.enabled() {
             EmitObj::Bitcode
-        } else if sess.opts.debugging_opts.embed_bitcode {
+        } else if sess.opts.debugging_opts.no_embed_bitcode {
+            EmitObj::ObjectCode(BitcodeSection::None)
+        } else {
             match sess.opts.optimize {
                 config::OptLevel::No | config::OptLevel::Less => {
-                    EmitObj::ObjectCode(BitcodeSection::Marker)
+                    // njn: I originally had BitcodeSection::Marker here, but
+                    // that causes a failure in
+                    // test/debuginfo/cross-crate-type-uniquing.rs, because
+                    // LLVMRustGetBitcodeSliceFromObjectData fails on an empty
+                    // section, because findBitcodeInObject() fails on an empty
+                    // section. Using `BitcodeSection::Full` here fixes things.
+                    //
+                    // More generally, I'd like to eliminate one of the
+                    // `BitcodeSection` variants anyway, as well as this
+                    // behaviour that depends on the optimization level, and
+                    // potential behaviour that depends on the target.
+                    //
+                    // Option 1: remove BitcodeSection::None, use
+                    // BitcodeSection::Marker in the `no_embed_bitcode` case
+                    // above, and use BitcodeSection::Full here. That way a
+                    // .o file containing object code always has a bitcode
+                    // section, which is full by default, and empty if
+                    // -Zno-embed-bitcode is given.
+                    //
+                    // Option 2: remove BitcodeSection::Marker. In theory it's
+                    // needed for iOS but it wasn't being used prior to this
+                    // PR. That way a .o file containing object code has a full
+                    // bitcode section by default, and no bitcode section if
+                    // -Zno-embed-bitcode is given.
```

Comment:
Ah ok I think I understand what's going on here. This is due to the fact that the `BitcodeSection` choice here didn't really matter before but now because it's the source of LTO bitcode it matters a whole lot more.

I believe that this is correct to, in all modes, include the full bitcode section. That's required for backwards-compatibility with rustc today where any rlib is always Rust-LTO-able.

This logic can likely switch to, as you mentioned in chat, switching to a marker unconditionally if `-Zembed-bitcode=no` is passed (or w/e the name of the switch is)

Reply:
We discussed this on Zulip and agreed to eliminate "marker" sections.